### PR TITLE
Add various combinators

### PR DESCRIPTION
**Pull request by pxeger**
Infix aliases:
- `@~` (`modify`)
- `@.` (`set`)
- `^.` (`get`)
- `^?` (`getOption`)
- `@?` (`l @? b = l @. Just b`)

And similar ones for setters in a state monad: `=~`, `=.`, `=?`

Others:
- `mapped` (a `setter` which focuses on every element in a `Functor`)
- `empty` (a `traversal` which doesn't focus on anything)
  - better name?
- homogeneous tuples can now be indexed
- `At`, and instance for `Map`s

### CODE REVIEW

**Comment by yallop** [comment]
It'd be nice to add support for [`Map`](https://github.com/ocaml/ocaml/blob/trunk/stdlib/map.mli) (whose type doesn't quite correspond to anything in Haskell).

Also: does mutable data fit into this approach?  e.g. could we add support for OCaml's arrays, references, hash tables, etc.?

**Comment by pxeger** [comment]
`Map` is on my radar - I was going to investigate @dvlasits' https://github.com/modular-implicits/datastructures to see if that was relevant first. I've also deliberately not added support for anything mutable*, to avoid having to think about the semantics too hard.

*(except `bytes` - but that was an accident)

File: tests/test.ml [hunk 1]
```diff
@@ -16,4 +16,55 @@ let () =
...
+  Any_Int.__any__
+
+let () =
+  let open Imp.Any in
+  let open Imp.Control in
+  let open Imp.Transformers in
+  let open Lens in
+  let module S = State {Any_Option {Any_Int}} in
+  let (>>) a b = bind {S} a (fun _ -> b) in
+  let t = traversed {Imp.Control.Option} in
```
Comment (pxeger) [inline tests/test.ml:61]: It's annoying that you can't pass something of type `{M: T} -> t` where something of type `t` is expected and have it infer `M`, even though you can workaround it by changing it to `{M: T} -> unit -> t`, and calling it with `()`, so it now just has type `t`.

Assuming suitable modules and module types:
```ocaml
let x {M: T} = M.x
let y (x: int) = x + 1
y x (* doesn't work *)


let x {M: T} () = M.x
let y (x: int) = x + 1
y (x ()) (* works *)
```

Comment (yallop) [reply tests/test.ml:61]: Yes, the need to pass `()` is sometimes less convenient.  The reason it's designed that way is that passing an implicit argument can cause some effects to occur (consider: `let f {M} = perform_effects (); v`), so we chose to allow implicit arguments to be instantiated only in explicit function calls, not in expressions that look like values.

**Pull request merged**